lib/quoil/cli.ex

```diff
@@ -0,0 +1,61 @@
+defmodule Quoil.CLI do
+  @moduledoc """
+  Command line interface for the quoil function  
+    -> parsing the arguments
+    -> running the ping command  
+    -> interpreting the results  
+    -> logging
+    """
+
+  def main(argv) do
+    argv
+    |> Quoil.ArgsProcessor.parse_args
+    |> terminate_early?
+    |> run_ping
+    |> Quoil.ParseResults.parse_result
+    |> Quoil.LogResults.write_log
+    |> Quoil.Repeat.requested?
+
+  end
+
+
+  def terminate_early?(:help) do
+    IO.puts """
+    usage: quoil [-h | --help]
+    quoil [--interval sec] [--number nr] [--repeat nr [--wait min]] <ip_to_ping> [log_file_name]
+    
+    all values supplied must be positive integers
+
+    For more information see: https://github.com/rawdamedia/quoil
+    """
+    System.halt(0)
+  end
+  def terminate_early?(:error) do
+    IO.puts "ERROR: There was an error with the command line switches."
```

Comment:
Using an error monad library means your functions don't see the `:ok` or `:error` element in the tuple, and just the payload. The monad effectively wraps metadata around the actual data you are working with, and handles wrapping, and unwrapping it.

In the case of an error, it **doesn't** call subsequent functions. It just returns the error straight away and bypasses them.

It's a more elegant way of writing deep nested logic to handle errors at any point.

So rather than writing

```elixir
case do_a_thing(data) do
  {:ok, payload} ->
    case do_another_thing(payload) do
      {:ok, another_payload} ->
        case do_yet_another_thing(another_payload) do
          {:ok, yet_another_payload} -> yet_another_payload
          {:error, error} -> handle_error(error)
        end
      {:error, error} -> handle_error(error)
    end
  {:error, error} -> handle error
end
```

you can do
```elixir
result = Error.p do
  Error.return(data)
  |> do_a_thing
  |> do_another_thing
  |> do_yet_another_thing
end

case result do
  {:error, error} -> handle_error(error)
  {:ok, value} -> value
end
```

It's less commonly used in Elixir than in something like Haskell, but it makes code like this much cleaner and easier to work with